src/tracker.rs

```diff
@@ -0,0 +1,247 @@
+                    action(metric, &tags, &seen);
+                }
+            }
+        }
+    }
+}
+
+/// Actions that define what the tracker will do when the custom metric threshold is passed.
+/// A user may define any number of these, and by default none are taken.
+enum ThresholdAction {
+    /// Emit an event. The count of unique tag sets, per metric, is provided as the tags
+    Event { title: String, text: String },
+
+    /// Take some custom action.
+    Custom(ThresholdCustomAction),
+}
+
+type ThresholdCustomAction = Box<dyn FnMut(&str, &[&str], &HashMap<String, Vec<HashSet<String>>>) + Send + Sync>;
+
+/// The configuration for the tag tracker. By default, the tag tracking is not enabled.
+/// To enable it, set the `count_threshold` to a non-zero value, and add at least one event
+/// or custom action
+/// Example usage:
+/// ```rust
+/// use prima_datadog::{
+///     configuration::{Country, PrimaConfiguration},
+///     Datadog, TagTrackerConfiguration,
+/// };
+/// let tracker_config = TagTrackerConfiguration::new()
+///     .with_threshold(21)
+///     .with_custom_action(|_, _, _| {});
+/// let configuration = PrimaConfiguration::new(
+///     "0.0.0.0:1234",
+///     "0.0.0.0:0",
+///     "prima_datadog_benchmarks",
+///     "dev".parse().unwrap(),
+/// ).with_country(Country::It).with_tracker_configuration(tracker_config);
+/// Datadog::init(configuration).unwrap();
+/// ```
+pub struct TagTrackerConfiguration {
+    count_threshold: usize,
+    actions: Vec<ThresholdAction>,
+}
+
+impl Default for TagTrackerConfiguration {
+    fn default() -> Self {
+        Self {
+            count_threshold: DEFAULT_TAG_THRESHOLD,
+            actions: Vec::new(),
+        }
+    }
+}
+
+impl TagTrackerConfiguration {
+    pub fn new() -> Self {
+        Self {
+            count_threshold: DEFAULT_TAG_THRESHOLD,
+            actions: Vec::new(),
+        }
+    }
+
+    /// Configure the tracker to emit an event when the threshold is reached.
+    /// The count of unique tag sets, per metric, is provided as the tags
+    /// for the event, i.e. for a given metric `metric`, there will be a tag `metric:count`,
+    /// where count is the number of unique tag sets seen for that metric.
+    /// Any number of events may be configured, and all will be emitted when the
+    /// threshold is reached.
+    pub fn with_event(mut self, title: String, text: String) -> Self {
+        self.actions.push(ThresholdAction::Event { title, text });
+        self
+    }
+
+    /// Add a custom action to execute when the custom metric threshold is reached.
+    /// These actions are run exactly once, at the point the threshold is reached.
+    /// Any number of actions may be added. The function will be passed the metric name
+    /// and tags of the metric causing the crossing of the threshold, and a
+    /// BTreeMap<String, Vec<BTreeSet<String>>> of all the unique tag sets seen for each metric.
```

Comment:
This only needs to be updated to reflect the latest changes
```suggestion
    /// Any number of actions may be added. The function will be passed the metric name,
    /// the tags of the metric causing the crossing of the threshold, and a
    /// HashMap containing all the unique tag sets seen for each metric.
```

Reply:
:pray: